pkg/collector/helm_collector.go

```diff
@@ -0,0 +1,88 @@
+package collector
+
+import (
+	"github.com/Azure/aks-periscope/pkg/interfaces"
+	"github.com/Azure/aks-periscope/pkg/utils"
+	"path/filepath"
+)
+
+// HelmCollector defines a Helm Collector struct
+type HelmCollector struct {
+	BaseCollector
+}
+
+var _ interfaces.Collector = &HelmCollector{}
+
+// NewHelmCollector is a constructor
+func NewHelmCollector(exporter interfaces.Exporter) *HelmCollector {
+	return &HelmCollector{
+		BaseCollector: BaseCollector{
+			collectorType: Helm,
+			exporter:      exporter,
+		},
+	}
+}
+
+// Collect implements the interface method
+func (collector *HelmCollector) Collect() error {
+	rootPath, err := utils.CreateCollectorDir(collector.GetName())
+	if err != nil {
+		return err
+	}
+
+	output, err := utils.RunCommandOnContainer("apk", "add", "curl", "openssl", "bash", "--no-cache")
+	if err != nil {
+		return err
+	}
+
+	output, err = utils.RunCommandOnContainer("curl", "-fsSl", "-o", "get_helm.sh", "https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-3")
+	if err != nil {
+		return err
+	}
+
+	output, err = utils.RunCommandOnContainer("chmod", "+x", "get_helm.sh")
+	if err != nil {
+		return err
+	}
+
+	output, err = utils.RunCommandOnContainer("./get_helm.sh")
+	if err != nil {
+		return err
+	}
+
+	helmListFile := filepath.Join(rootPath, "helm_list")
+	output, err = utils.RunCommandOnContainer("helm", "list", "--all-namespaces")
```

Comment:
🐛 More details here: https://github.com/Azure/aks-periscope/pull/54#discussion_r634838600

Reply:
💡 Let's just dig a little deeper for this: I think adding `secrets` to the list might be ok, and non-impacting, as this all runs with users permission and user in use will run it and that might enable the `Arc` - `helm` scenarios as well.

Reply:
Lets add the `secrets` in resources and keep 1 yaml, I am not able to find out any issues with it, inc are we did we can always have a fallback idea anyways. Thank you. here is the small PR for it: https://github.com/sophsoph321/aks-periscope/pull/1 (Mainly to simplify all this)

Reply:
This looks like a *very* risky set of permissions to me? This creates a ClusterRole with list permissions for secrets across all namespaces in the cluster. 

What is this needed for? Is there any way we can scope the permissions down to e.g. enable secret get to just that secret which is required?

e.g. number 1 risky permission here: https://www.cyberark.com/resources/threat-research-blog/securing-kubernetes-clusters-by-eliminating-risky-permissions
https://darkbit.io/blog/the-power-of-kubernetes-rbac-list

Reply:
@davidkydd  The reasoning comes from here: https://helm.sh/docs/topics/rbac/#example-grant-a-user-read-only-access-at-the-cluster-scope 
If you go all the way at the bottom of that link above, it says in order to run "helm list --all-namespaces", the user needs cluster-scope read access, which requires having both view and secret-reader access. I discussed this internally with arc team, and we couldn't figure out a different way to work around the issues Tats was having with running "helm list".

Reply:
Yeah, +1 to this, essentially this needs to be explored, it feels like helm commands are restrictive this way, but Worth a discussion. (Wider arc discussion)

I would prefer this to be pure kubectl driven: 

* kubectl -n azure-arc get deployments,pods 

Above command will give much better, but if he’ll command I 100% needed we need to make sure we cover these small things and rationales it well.

Reply:
For kubectl driven we already have KubeObject to get deployments, pods. This is specific with Helm as it is capturing helm history and its revisions. We discussed about cluster scope access specific to Arc Scenarios earlier (I think this was in terms of CRD access).  Two options that were discussed:

1. Update Periscope clusterRole to include additional permissions
https://github.com/Azure/aks-periscope/blob/master/deployment/aks-periscope.yaml#L12

2. Instead of updating clusterRole (so as to stick to be least privilege for AKS), The CLI will update the rolebinding after periscope deployment to provide additional access to service account

I would have preferred option 1, provided we take consent from user. We can use option 2 for time being if 1 is no go.

Reply:
Yeah, I was thinking if we are keeping this way then we can keep current PR level implementation as it is and with the privacy policy along with case that user need to run this tool.

https://github.com/Azure/aks-periscope#data-privacy-and-collection

Reply:
Sure, we will have data privacy link

Reply:
@davidkydd, Can you share some more context on security concern on more permission with periscope? Is it a generic comment or some issue which periscope encountered in past. Since periscope gets deleted once the log collection is done, wanted to understand the security concerns in detail. 

Like I mentioned above we have two choices, given by default in clusterrole def or given CLI after installing periscope. I do not see much difference between two as  permission would be need to be given in either case by user on cluster to run the helm list cmd (Probably for CRDs as well in future, once we reach to this collector). More details on this will help on the choice on how and by which component to enable these role permissions.   

My assumption today is periscope is already letting user know what is going to do on the cluster with consent, hence asking the addition role permission concerns in detail.

Reply:
not due to something experienced in the past , just jumped out as a red flag. Also note that while CLI deployment of periscope will be cleaned up once complete, non-CLI driven deployments aren't cleaned up automatically, and would potentially leave this dangling clusterRole.

Assumed you would be able to at least scope the secrets you need to read down so there wouldn't be need for full "list" permission on a clusterRole, but from the link @sophsoph321 posted it doesn't look like it. If this is already at least-privilege possible for the commands the helm collector needs to run then think thats probably fine.

As we will likely encounter with the CRD permissions - it seems we are forced to either:

1) introduce additional complexity (Arc CLI has to update roleBinding after deploy), or 
2) violate least privilege by giving all flavours of deployments the full super-set of required permissions (AKS deployments of periscope will end up with CRD + secrets list permissions when it makes no use of them)

Again, a problem which could be avoided by using a different config / deployment / yaml file for each integration

Reply:
Yeah, initial thought process to be careful was exactly what I thought, but essentially the use case for this tool will not have any security impact to the behaviour.

User will always run this tool on their known cluster context. I would see it as Helm enablement requirement since this issue common with `Helm` command. 

To get more idea/opinion, I reached out to @JunSun17 (**Thank you** for reply) - as long as we are not logging any secrets in an exported file. `I think the tool can access secrets and use it in the program to perform some logic, just make sure the secret values are not written out when exporting, or it will be exposed.` 

Further, @safeermohammed kindly agrees to stay involved and if there is any need be `Arc` will own and we can together make move forward.

Side note: I would say, we can go ahead given we have captured all the opinion and have plan-b in case we find any concrete scenario. Also, if this at all is any security bug in `Helm` then it will be `helm` issue and in that case I would rather prefer to disable the whole `helm` command.

Thanks guys. 🙏

Reply:
Thanks @Tatsinnit - my concern was mostly driven by my own naivety re: k8s security :) 

Believe this can be resolved if nothing else from Arc side

Reply:
@Tatsinnit thanks for summarizing this! I totally agree with you on the security side.

Reply:
As a quick note regarding "just make sure the secret values are not written out when exporting, or it will be exposed." Secrets *will* be exported out, but only those corresponding to helm releases, which is by design